Turn the Package model into a single-table inheritance model

jeremycline opened the pull request:
This turns the Package model into the parent class using single-table
inheritance that sub-classes can inherit from to create different
Package types. It includes a migration that sets the type of all
existing Packages to RpmPackage.

fixes #1392

jeremycline committed:
Turn the Package model into a single-table inheritance model

This turns the Package model into the parent class using single-table
inheritance that sub-classes can inherit from to create different
Package types. It includes a migration that sets the type of all
existing Packages to RpmPackage.

fixes #1392

Signed-off-by: Jeremy Cline <jeremy@jcline.org>

jeremycline committed:
Change all uses of Package to RpmPackage

Signed-off-by: Jeremy Cline <jeremy@jcline.org>

bowlofeggs commented on alembic/versions/fc6b0169c596_add_a_column_for_single_table_.py:
```diff
@@ -0,0 +1,26 @@
+def upgrade():
+    """Add the column used for polymorphic identity in Packages."""
+    # The default of ``1`` is the RPM package type.
+    op.add_column('packages', sa.Column('type', sa.Integer(), nullable=False, default=1))
```
This migration fails to apply:

```
[vagrant@bodhi-dev bodhi]$ alembic upgrade head
/usr/lib64/python2.7/pkgutil.py:186: ImportWarning: Not importing directory '/usr/lib64/python2.7/site-packages/zope': missing __init__.py
  file, filename, etc = imp.find_module(subname, path)
/usr/lib64/python2.7/pkgutil.py:186: ImportWarning: Not importing directory '/usr/lib/python2.7/site-packages/zope': missing __init__.py
  file, filename, etc = imp.find_module(subname, path)
/usr/lib64/python2.7/pkgutil.py:186: ImportWarning: Not importing directory '/usr/lib/python2.7/site-packages/paste': missing __init__.py
  file, filename, etc = imp.find_module(subname, path)
/usr/lib64/python2.7/pkgutil.py:186: ImportWarning: Not importing directory '/usr/lib/python2.7/site-packages/moksha': missing __init__.py
  file, filename, etc = imp.find_module(subname, path)
/usr/lib64/python2.7/pkgutil.py:186: ImportWarning: Not importing directory '/usr/lib/python2.7/site-packages/repoze': missing __init__.py
  file, filename, etc = imp.find_module(subname, path)
/home/vagrant/bodhi/bodhi/server/__init__.py:26: DeprecationWarning: unauthenticated_userid: As of Pyramid 1.5 the "pyramid.security.unauthenticated_userid" API is now deprecated.  It will be removed in Pyramd 1.8.  Use the "unauthenticat
ed_userid" attribute of the Pyramid request instead.
  from pyramid.security import unauthenticated_userid
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade 4f2f825bcf4a -> fc6b0169c596, Add a column for single-table inheritance to the packages table.
Traceback (most recent call last):
  File "/usr/bin/alembic", line 9, in <module>
    load_entry_point('alembic==0.8.6', 'console_scripts', 'alembic')()
  File "/usr/lib/python2.7/site-packages/alembic/config.py", line 479, in main
    CommandLine(prog=prog).main(argv=argv)
  File "/usr/lib/python2.7/site-packages/alembic/config.py", line 473, in main
    self.run_cmd(cfg, options)
  File "/usr/lib/python2.7/site-packages/alembic/config.py", line 456, in run_cmd
    **dict((k, getattr(options, k)) for k in kwarg)
  File "/usr/lib/python2.7/site-packages/alembic/command.py", line 174, in upgrade
    script.run_env()
  File "/usr/lib/python2.7/site-packages/alembic/script/base.py", line 397, in run_env
    util.load_python_file(self.dir, 'env.py')
  File "/usr/lib/python2.7/site-packages/alembic/util/pyfiles.py", line 93, in load_python_file
    module = load_module_py(module_id, path)
  File "/usr/lib/python2.7/site-packages/alembic/util/compat.py", line 79, in load_module_py
    mod = imp.load_source(module_id, path, fp)
  File "alembic/env.py", line 75, in <module>
    run_migrations_online()
  File "alembic/env.py", line 68, in run_migrations_online
    context.run_migrations()
  File "<string>", line 8, in run_migrations
  File "/usr/lib/python2.7/site-packages/alembic/runtime/environment.py", line 797, in run_migrations
    self.get_context().run_migrations(**kw)
  File "/usr/lib/python2.7/site-packages/alembic/runtime/migration.py", line 312, in run_migrations
    step.migration_fn(**kw)
  File "/home/vagrant/bodhi/alembic/versions/fc6b0169c596_add_a_column_for_single_table_.py", line 20, in upgrade
    op.add_column('packages', sa.Column('type', sa.Integer(), nullable=False, default=1))
  File "<string>", line 8, in add_column
  File "<string>", line 3, in add_column
  File "/usr/lib/python2.7/site-packages/alembic/operations/ops.py", line 1535, in add_column
    return operations.invoke(op)
  File "/usr/lib/python2.7/site-packages/alembic/operations/base.py", line 318, in invoke
    return fn(self, operation)
  File "/usr/lib/python2.7/site-packages/alembic/operations/toimpl.py", line 123, in add_column
    schema=schema
  File "/usr/lib/python2.7/site-packages/alembic/ddl/impl.py", line 172, in add_column
    self._exec(base.AddColumn(table_name, column, schema=schema))
  File "/usr/lib/python2.7/site-packages/alembic/ddl/impl.py", line 118, in _exec
    return conn.execute(construct, *multiparams, **params)
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/engine/base.py", line 914, in execute
    return meth(self, multiparams, params)
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/sql/ddl.py", line 68, in _execute_on_connection
    return connection._execute_ddl(self, multiparams, params)
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/engine/base.py", line 968, in _execute_ddl
    compiled
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/engine/base.py", line 1146, in _execute_context
    context)
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/engine/base.py", line 1341, in _handle_dbapi_exception
    exc_info
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/util/compat.py", line 202, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/engine/base.py", line 1139, in _execute_context
    context)
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/engine/default.py", line 450, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.IntegrityError: (psycopg2.IntegrityError) column "type" contains null values
 [SQL: 'ALTER TABLE packages ADD COLUMN type INTEGER NOT NULL']
```

I think we need to make it nullable, then run a query to set existing rows to ```1```, then run an update to make it non-nullable.

jeremycline replied:
The setting I wanted was ``server_default`` rather than ``default``. This lets you immediately set the column to be non-nullable (because it'll use the default for everything) and then you just alter the column to drop the server default. It comes to the same thing, though.

bowlofeggs commented on bodhi/server/models.py:
```diff
@@ -476,12 +476,28 @@ class TestCase(Base):
+
+    Attributes:
+        name (unicode): A unicode string that uniquely identifies the package.
+        requirements (unicode): A unicode string that describes the requirements of a package.
```
Perhaps "A unicode string that lists space-separated taskotron test results that must pass for this package."

bowlofeggs commented on bodhi/server/models.py:
```diff
@@ -476,12 +476,28 @@ class TestCase(Base):
+            class to create when loading rows from the database.
+        builds (list): A list of :class:`Build` objects.
+        test_cases (list): A list of :class:`TestCase` objects.
+        committers (list): A list of :class:`User` objects who are committers.
```
Are these really lists, or are they generators or some sqlalchemy type?

jeremycline replied:
I thought they were just lists because I checked on Anitya, but it turned out that was a property making them a list. It's actually an SQLAlchemy type. I updated the docs.

jeremycline committed:
Use server_default rather than default

The migration was failing since the correct argument to set the server's
default value for a column is ``server_default``.

Signed-off-by: Jeremy Cline <jeremy@jcline.org>

jeremycline committed:
Update the documentation on the Package model

Signed-off-by: Jeremy Cline <jeremy@jcline.org>

bowlofeggs reviewed (approved):
LGTM.
